# Shwet Jain

# IST 707 - Assignment 3 - Naive Bayes , SVM , Random Forest

# Section 1:

# Introduction

The dataset consists of the pixel values in gray scale. This dataset stores the pixel values for different digits from 0 to 9. This train dataset will be used for training the model and predicting the digits from the pixel values of the test dataset. This homework involves training data using different methods such as Naive Bayes, kNN and SVM models. By comparing the accuracies of these models, prediction would be done on the test data using the model with the highest accuracy

## Data Loading and Pre-Processing 

Importing all the necessary packages

In [496]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.model_selection import train_test_split, GridSearchCV, cross_val_score, ShuffleSplit
from sklearn import metrics
from sklearn.naive_bayes import GaussianNB, BernoulliNB, MultinomialNB

Loading the sample dataset

In [497]:
train_data = pd.read_csv('/Users/shwet/Downloads/Kaggle-digit-train-sample-small-1400.csv')
test_data = pd.read_csv('/Users/shwet/Downloads/Kaggle-digit-test-sample1000.csv')
train_data.shape

(1400, 785)

Dropping the first column of the loaded dataset and storing it in X

In [498]:
X = train_data.drop('label', axis=1)

### Data Preprocessing for Naive Bayesian

Pixels greater than 128 have been replaced by 1 and less than 128 have been replaced by 0 so that variance in the model is reduced and well explained. (This is not necessary though)


In [499]:
for x in X.columns[0:]:
    X[x]=np.where(X[x] <= 128, 1,0)

Storing the first column in y 

In [500]:
y = train_data['label'] 

Splitting data into train data(70%) and test data(30%) using train_test_split

In [501]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=16)

Importing the time function

In [502]:
from time import time

# Section 2: 

## Naive Bayes Algorithm

Importing the MultinomialNB function

In [503]:
from sklearn.naive_bayes import MultinomialNB

Running the model with default parameters

In [504]:
gnb = MultinomialNB()
start = time()
gnb_param={}
grid_nb=GridSearchCV(gnb, gnb_param,cv=5,scoring="accuracy")
grid_nb.fit(X_train, y_train)
print(" Naive Bayesian took %.2f seconds"
       % (time() - start))

 Naive Bayesian took 0.11 seconds


In [505]:
grid_nb.best_score_

0.8051020408163265

In [506]:
gnb_pred = grid_nb.predict(X_test)
acc= round(metrics.accuracy_score(y_test, gnb_pred)*100, 2)
print ("Accuracy: " + str(acc) + "%")

Accuracy: 80.48%


Naive Bayesian takes in one parameter alpha, hence fine tuning the model using Grid Search 

In [507]:
param_grid= {'alpha':[0,0.2,0.5,1]}

In [508]:
start = time()
grid_nb = GridSearchCV(gnb, param_grid, cv=10, scoring='accuracy')
grid_nb.fit(X_train, y_train)
print(" Naive Bayesian took %.2f seconds"
       % (time() - start))


C:\Users\shwet\Anaconda2\lib\site-packages\sklearn\naive_bayes.py:480: UserWarning: alpha too small will result in numeric errors, setting alpha = 1.0e-10
  'setting alpha = %.1e' % _ALPHA_MIN)
C:\Users\shwet\Anaconda2\lib\site-packages\sklearn\naive_bayes.py:480: UserWarning: alpha too small will result in numeric errors, setting alpha = 1.0e-10
  'setting alpha = %.1e' % _ALPHA_MIN)
C:\Users\shwet\Anaconda2\lib\site-packages\sklearn\naive_bayes.py:480: UserWarning: alpha too small will result in numeric errors, setting alpha = 1.0e-10
  'setting alpha = %.1e' % _ALPHA_MIN)
C:\Users\shwet\Anaconda2\lib\site-packages\sklearn\naive_bayes.py:480: UserWarning: alpha too small will result in numeric errors, setting alpha = 1.0e-10
  'setting alpha = %.1e' % _ALPHA_MIN)
C:\Users\shwet\Anaconda2\lib\site-packages\sklearn\naive_bayes.py:480: UserWarning: alpha too small will result in numeric errors, setting alpha = 1.0e-10
  'setting alpha = %.1e' % _ALPHA_MIN)
C:\Users\shwet\Anaconda2\lib\s

 Naive Bayesian took 0.82 seconds


C:\Users\shwet\Anaconda2\lib\site-packages\sklearn\model_selection\_search.py:841: DeprecationWarning: The default of the `iid` parameter will change from True to False in version 0.22 and will be removed in 0.24. This will change numeric results when test-set sizes are unequal.
  DeprecationWarning)


To check the best parameter amongst the provided values in the list 

In [510]:
grid_nb.best_params_

{'alpha': 0.5}

Testing the model on the test data 

In [511]:
grid_pred = grid_nb.predict(X_test)
acc= round(metrics.accuracy_score(y_test, grid_pred)*100, 2)
print ("Accuracy: " + str(acc) + "%")

Accuracy: 80.48%


### Result: Naive Bayesian - Time taken : 0.82 secs , Accuracy - 80.48%

# Section 3 - Part 1

# Linear SVM

In [534]:
train_data = pd.read_csv('/Users/shwet/Downloads/Kaggle-digit-train-sample-small-1400.csv')
test_data = pd.read_csv('/Users/shwet/Downloads/Kaggle-digit-test-sample1000.csv')

In [535]:
X = train_data.drop('label', axis=1)

In [536]:
y = train_data['label'] 

In [537]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=16)

import all the functions and packages

In [518]:
from sklearn import svm
from sklearn.svm import SVC

In [519]:
lin_clf = svm.LinearSVC(max_iter=100000)

Passed in different values of C for Grid Search

In [520]:
param_grid = {'C':[0.1,0.25,0.5,0.75,1]}


In [521]:
start = time()
grid_lsvm = GridSearchCV(lin_clf,param_grid,cv=5,scoring="accuracy")
grid_lsvm.fit(X_train,y_train)
print("Linear SVM took %.2f seconds"
       % (time() - start))

Linear SVM took 8.12 seconds


In [522]:
grid_lsvm.best_params_

{'C': 0.1}

In [523]:
grid_pred = grid_lsvm.predict(X_test_std)
acc= round(metrics.accuracy_score(y_test, grid_pred)*100, 2)
print ("Accuracy: " + str(acc) + "%")

Accuracy: 77.62%


### Result: Linear SVM - Time taken : 8.12 secs , Accuracy - 77.62%

# Section 3 - part 2 

# Non-Linear SVM to increase accuracy of the model 

In [524]:
tuned_parameters = {'kernel': ['rbf','poly','sigmoid'],'degree':[2,3,5], 'gamma':['scale'],
                     'C':[0.01,0.4,0.8,1]}

In [525]:
start = time()
svml = svm.SVC()                    
grid_nsvm = GridSearchCV(svml,tuned_parameters,cv=5, scoring='accuracy')
grid_nsvm.fit(X_train, y_train)
print("Linear SVM took %.2f seconds"
       % (time() - start))

C:\Users\shwet\Anaconda2\lib\site-packages\sklearn\model_selection\_search.py:841: DeprecationWarning: The default of the `iid` parameter will change from True to False in version 0.22 and will be removed in 0.24. This will change numeric results when test-set sizes are unequal.
  DeprecationWarning)


Linear SVM took 256.02 seconds


In [543]:
grid_nsvm.best_params_

{'C': 1, 'degree': 2, 'gamma': 'scale', 'kernel': 'rbf'}

In [528]:
grid_pred = grid_nsvm.predict(X_test)
acc= round(metrics.accuracy_score(y_test, grid_pred)*100, 2)
print ("Accuracy: " + str(acc) + "%")

Accuracy: 90.48%


### Result: Non-Linear SVM - Time taken : 256.02 secs , Accuracy - 90.48%

# Section 4 :

# Random Forest Classifier

Importing RandomForestClassifier function

In [538]:
from sklearn.ensemble import RandomForestClassifier

Setting 'param_drid' for Grid Search

In [539]:
 param_grid = {"max_depth": [3,7,10,5,15], "n_estimators":[50,100]}


Running the model 

In [540]:
rfc= RandomForestClassifier()
start = time()
grid_rfc = GridSearchCV(rfc, param_grid, cv=5, scoring='accuracy')
grid_rfc.fit(X_train, y_train)
print(" Random Forest Classifier %.2f seconds"
       % (time() - start))

C:\Users\shwet\Anaconda2\lib\site-packages\sklearn\model_selection\_search.py:841: DeprecationWarning: The default of the `iid` parameter will change from True to False in version 0.22 and will be removed in 0.24. This will change numeric results when test-set sizes are unequal.
  DeprecationWarning)


 Random Forest Classifier 14.91 seconds


In [541]:
grid_rfc.best_params_

{'max_depth': 7, 'n_estimators': 100}

Testing the model on test data 

In [542]:
grid_pred = grid_rfc.predict(X_test)
acc= round(metrics.accuracy_score(y_test, grid_pred)*100, 2)
print ("Accuracy: " + str(acc) + "%")

Accuracy: 87.14%


### Result: Random Forest Classifier - Time taken : 14.91 secs , Accuracy - 87.14%

# Section 5:

# Conclusion

### Out of the three model, Naive Bayesian is the fastest (0.82 sec) but has low accuracy(80.48%) in comparison to other two models 
 

### On the other hand, SVM has an accuracy of 90.48% and it takes 256.02 secs

### Random Forest Classifier has an accuracy of 87.14% nearly to that of SVM but it takes comparatively less time than SVM i.e. it takes 14.91 secs

## SVM has the highest accuracy

### Therefore, we should run our test data on SVM 

In [544]:
testx = test_data.drop('label',axis=1)
testx = testx.iloc[0:998 :]
grid_pred = grid_nsvm.predict(testx)


In [545]:
grid_pred

array([2, 0, 9, 4, 2, 7, 0, 3, 0, 3, 5, 7, 4, 0, 4, 5, 3, 1, 9, 0, 9, 1,
       1, 5, 7, 4, 2, 7, 7, 7, 7, 5, 4, 2, 6, 2, 5, 5, 1, 6, 7, 7, 4, 9,
       5, 7, 8, 8, 6, 7, 6, 8, 8, 3, 8, 2, 1, 2, 9, 0, 4, 1, 7, 0, 0, 0,
       1, 4, 0, 1, 6, 5, 8, 8, 2, 8, 3, 9, 2, 8, 5, 9, 1, 0, 9, 2, 4, 3,
       6, 7, 2, 0, 6, 6, 1, 4, 3, 8, 7, 4, 0, 3, 2, 0, 7, 3, 0, 5, 0, 3,
       0, 0, 4, 7, 1, 7, 1, 1, 3, 3, 3, 7, 2, 8, 6, 3, 8, 7, 8, 4, 3, 5,
       6, 0, 0, 0, 3, 1, 3, 0, 4, 3, 4, 5, 5, 8, 7, 4, 2, 8, 4, 3, 5, 6,
       5, 1, 7, 5, 7, 8, 3, 0, 4, 5, 1, 8, 7, 6, 3, 0, 2, 7, 9, 6, 1, 3,
       7, 4, 1, 2, 4, 3, 5, 2, 4, 9, 2, 1, 6, 0, 6, 1, 4, 9, 6, 0, 9, 7,
       6, 9, 1, 9, 0, 9, 9, 0, 8, 4, 6, 2, 0, 9, 3, 6, 7, 2, 1, 6, 3, 4,
       2, 3, 1, 2, 2, 0, 4, 6, 1, 0, 0, 4, 9, 1, 7, 3, 2, 8, 8, 6, 8, 6,
       2, 8, 5, 5, 9, 8, 3, 8, 9, 7, 1, 3, 8, 4, 5, 1, 4, 5, 6, 3, 3, 5,
       7, 0, 6, 8, 6, 1, 6, 0, 6, 3, 9, 9, 1, 5, 8, 4, 0, 9, 2, 0, 5, 3,
       7, 1, 9, 9, 5, 7, 7, 9, 9, 6, 3, 0, 3, 3, 6,

# We have predicted the test data using SVM